core/shared_queue.js

```diff
@@ -151,14 +151,27 @@ SharedQueue Binary Layout
 
   function handleAsyncMsgFromRust(buf) {
     if (buf) {
-      asyncHandler(buf);
+      handleAsyncMsgFromRustInner(buf);
     } else {
       while ((buf = shift()) != null) {
-        asyncHandler(buf);
+        handleAsyncMsgFromRustInner(buf);
       }
     }
   }
 
+  function handleAsyncMsgFromRustInner(buf) {
+    // DataView to extract cmdId value.
+    const dataView = new DataView(buf.buffer, buf.byteOffset, 4);
+    const promiseId = dataView.getInt32(0);
```

Comment:
Ah... This is still bothering me because the buf is not opaque to core. This still defines some byte layout in the buffers.

The abstraction I want to make in //core is that these buffers are completely defined and controlled by the embedder and that core only passes them around.

Fixing this would require passing the promiseId as a parameter to the callback...

Reply:
...and the best way to achieve this when using the shared buffer to pass results is to put this data into the record. I can't really think of any better way to do this off of the top of my head, but maybe it makes more sense to decode this information from the record in `shift()`.

Reply:
Oh! So this is only on the shared buffer? I think I might not have understood this correctly.

As far as the interfaces in core/isolate.rs, are the buffers passed to ops opaque?

Reply:
Yes promise_id is just appended to the beginning of a records when they are added to the shared buffer via `push()`. If you look down another line you can see we shrink the window used for the returned buffer, so that the promise_id in the record  is not directly exposed to non-core js code.

Reply:
👍